kaggle_environments/envs/halite/helpers.py

```diff
@@ -0,0 +1,781 @@
+            # We'll populate the player's ships and shipyards in _add_ship and _add_shipyard
+            self.players[player_id] = Player(player_id, player_halite, [], [], self)
+            player_actions = next_actions[player_id] or {}
+
+            for (ship_id, [ship_index, ship_halite]) in player_ships.items():
+                # In the raw observation, halite is stored as a 1d list but we convert it to a 2d dict for convenience
+                # Accordingly we also need to convert our list indices to dict keys / 2d positions
+                ship_position = index_to_position(ship_index, size)
+                raw_action = player_actions.get(ship_id)
+                action = (
+                    ShipAction[raw_action]
+                    if raw_action is not None
+                    else None
+                )
+                self._add_ship(Ship(ship_id, ship_position, ship_halite, player_id, self, action))
+
+            for (shipyard_id, shipyard_index) in player_shipyards.items():
+                shipyard_position = index_to_position(shipyard_index, size)
+                raw_action = player_actions.get(shipyard_id)
+                action = (
+                    ShipyardAction[raw_action]
+                    if raw_action is not None
+                    else None
+                )
+                self._add_shipyard(Shipyard(shipyard_id, shipyard_position, player_id, self, action))
+
+    @property
+    def configuration(self) -> Configuration:
+        return self._configuration
+
+    @property
+    def players(self) -> Dict[PlayerId, Player]:
+        return self._players
+
+    @property
+    def ships(self) -> Dict[ShipId, Ship]:
+        """Returns all ships on the current board."""
+        return self._ships
+
+    @property
+    def shipyards(self) -> Dict[ShipyardId, Shipyard]:
+        """Returns all shipyards on the current board."""
+        return self._shipyards
+
+    @property
+    def cells(self) -> Dict[Point, Cell]:
+        """Returns all cells on the current board."""
+        return self._cells
+
+    @property
+    def step(self) -> int:
+        return self._step
+
+    @property
+    def current_player_id(self) -> PlayerId:
+        return self._current_player_id
+
+    @property
+    def current_player(self) -> Player:
+        """Returns the current player (generally this is you)."""
+        return self._players[self.current_player_id]
+
+    @property
+    def opponents(self) -> List[Player]:
+        """
+        Returns all players that aren't the current player.
+        You can get all opponent ships with [ship for ship in player.ships for player in board.opponents]
+        """
+        return [player for player in self.players.values() if not player.is_current_player]
+
+    @property
+    def observation(self) -> Dict[str, Any]:
+        """Converts a Board back to the normalized observation that constructed it."""
```

Comment:
This is the observation serializer to the constructor's observation deserializer